# Hackathon 2026 - Previsão Climática de Precipitação sobre a América do Sul

## Integrantes

* Bruno Antico Galin (@brunoantico)
* Lucas Pires de Camargo Sarai (@lsarai)
* Victor Luiz de Sá Alves (@viktoralves)


## Analisando dados disponibilizados

### Dados de Treino

In [1]:
%pip install tensorflow keras-tuner

Note: you may need to restart the kernel to use updated packages.


In [ ]:
import xarray as xr
import pandas as pd
import kagglehub
from __future__ import annotations

from pathlib import Path
from typing import Iterable

import numpy as np
import tensorflow as tf

def mostrar_infos_df(df: pd.DataFrame, label: str):
    print("#"*5,label.upper(),"#"*5)
    print(df.info())
    display(df.head())
    print("#"*20,"\n")

def contar_valores_por_time(df: pd.DataFrame) -> pd.DataFrame:
    if "time" not in df.index.names:
        raise ValueError("O DataFrame precisa ter `time` como índice.")

    valores = df.groupby(level="time").count()
    valores.index.name = "time"
    return valores

def contar_linhas_por_time(df: pd.DataFrame) -> pd.Series:
    return df.groupby(level="time").size().rename("total_registros")

def contar_dimensoes_grade(df: pd.DataFrame) -> dict[str, pd.Series]:
    nomes = set(df.index.names)
    esperados = {"time", "lat", "lon"}
    ausentes = esperados - nomes
    if ausentes:
        raise ValueError(f"O índice precisa conter: {sorted(esperados)}. Ausentes: {sorted(ausentes)}")

    coordenadas = df.index.to_frame(index=False)
    return {
        "latitudes_por_time": coordenadas.groupby("time")["lat"].nunique(),
        "longitudes_por_latitude": coordenadas.groupby("lat")["lon"].nunique(),
        "longitudes_por_time": coordenadas.groupby("time")["lon"].count(),
    }


In [3]:
path = kagglehub.competition_download('previsao-climatica-de-precipitacao-sobre-a-america-do-sul')

print("Path to competition files:", path)

Path to competition files: /kaggle/input/competitions/previsao-climatica-de-precipitacao-sobre-a-america-do-sul


In [4]:
print("#"*5,"TREINO","#"*5,"\n")
tp = xr.open_dataset(f"{path}/treino_tp.nc").to_dataframe()              # (time, lat, lon), mm/dia


mostrar_infos_df(tp,"precipitação mes atual")
display(contar_linhas_por_time(tp))
display(contar_valores_por_time(tp))
display(contar_dimensoes_grade(tp))

print(tp.index.get_level_values("lat").unique())
print(tp.index.get_level_values("lon").unique())



##### TREINO ##### 



##### PRECIPITAÇÃO MES ATUAL #####


<class 'pandas.core.frame.DataFrame'>
MultiIndex: 78246756 entries, (Timestamp('1940-01-01 00:00:00'), np.float64(-60.0), np.float64(-90.0)) to (Timestamp('2022-12-01 00:00:00'), np.float64(15.0), np.float64(-25.0))
Data columns (total 1 columns):
 #   Column  Dtype  
---  ------  -----  
 0   tp      float32
dtypes: float32(1)
memory usage: 746.3 MB
None


tp
time       lat   lon             
1940-01-01 -60.0 -90.00  2.254486
                 -89.75  2.285004
                 -89.50  2.338409
                 -89.25  2.424240
                 -89.00  2.477646

#################### 



time
1940-01-01    78561
1940-02-01    78561
1940-03-01    78561
1940-04-01    78561
1940-05-01    78561
              ...  
2022-08-01    78561
2022-09-01    78561
2022-10-01    78561
2022-11-01    78561
2022-12-01    78561
Name: total_registros, Length: 996, dtype: int64

,tp
time,
1940-01-01,78561
1940-02-01,78561
1940-03-01,78561
1940-04-01,78561
1940-05-01,78561
...,...
2022-08-01,78561
2022-09-01,78561
2022-10-01,78561


{'latitudes_por_time': time
 1940-01-01    301
 1940-02-01    301
 1940-03-01    301
 1940-04-01    301
 1940-05-01    301
              ... 
 2022-08-01    301
 2022-09-01    301
 2022-10-01    301
 2022-11-01    301
 2022-12-01    301
 Name: lat, Length: 996, dtype: int64,
 'longitudes_por_latitude': lat
 -60.00    261
 -59.75    261
 -59.50    261
 -59.25    261
 -59.00    261
          ... 
  14.00    261
  14.25    261
  14.50    261
  14.75    261
  15.00    261
 Name: lon, Length: 301, dtype: int64,
 'longitudes_por_time': time
 1940-01-01    78561
 1940-02-01    78561
 1940-03-01    78561
 1940-04-01    78561
 1940-05-01    78561
               ...  
 2022-08-01    78561
 2022-09-01    78561
 2022-10-01    78561
 2022-11-01    78561
 2022-12-01    78561
 Name: lon, Length: 996, dtype: int64}

Index([ -60.0, -59.75,  -59.5, -59.25,  -59.0, -58.75,  -58.5, -58.25,  -58.0,
       -57.75,
       ...
        12.75,   13.0,  13.25,   13.5,  13.75,   14.0,  14.25,   14.5,  14.75,
         15.0],
      dtype='float64', name='lat', length=301)


Index([ -90.0, -89.75,  -89.5, -89.25,  -89.0, -88.75,  -88.5, -88.25,  -88.0,
       -87.75,
       ...
       -27.25,  -27.0, -26.75,  -26.5, -26.25,  -26.0, -25.75,  -25.5, -25.25,
        -25.0],
      dtype='float64', name='lon', length=261)


In [5]:
alvo = xr.open_dataset(f"{path}/treino_tp_alvo.nc").to_dataframe()   # precipitação de M+1

mostrar_infos_df(alvo,"precipitação mes seguinte")
display(contar_linhas_por_time(alvo))
display(contar_valores_por_time(alvo))

##### PRECIPITAÇÃO MES SEGUINTE #####


<class 'pandas.core.frame.DataFrame'>
MultiIndex: 78246756 entries, (Timestamp('1940-01-01 00:00:00'), np.float64(-60.0), np.float64(-90.0)) to (Timestamp('2022-12-01 00:00:00'), np.float64(15.0), np.float64(-25.0))
Data columns (total 1 columns):
 #   Column   Dtype  
---  ------   -----  
 0   tp_alvo  float32
dtypes: float32(1)
memory usage: 746.3 MB
None


tp_alvo
time       lat   lon             
1940-01-01 -60.0 -90.00  3.153801
                 -89.75  3.170013
                 -89.50  3.149986
                 -89.25  3.075600
                 -89.00  2.992630

#################### 



time
1940-01-01    78561
1940-02-01    78561
1940-03-01    78561
1940-04-01    78561
1940-05-01    78561
              ...  
2022-08-01    78561
2022-09-01    78561
2022-10-01    78561
2022-11-01    78561
2022-12-01    78561
Name: total_registros, Length: 996, dtype: int64

,tp_alvo
time,
1940-01-01,78561
1940-02-01,78561
1940-03-01,78561
1940-04-01,78561
1940-05-01,78561
...,...
2022-08-01,78561
2022-09-01,78561
2022-10-01,78561


In [6]:
cobertura_nuvem = xr.open_dataset(f"{path}/treino_cloud_cover.nc").to_dataframe()

mostrar_infos_df(cobertura_nuvem,"cobertura nuvem")
display(contar_valores_por_time(cobertura_nuvem))
display(contar_linhas_por_time(cobertura_nuvem))

##### COBERTURA NUVEM #####


<class 'pandas.core.frame.DataFrame'>
MultiIndex: 78246756 entries, (Timestamp('1940-01-01 00:00:00'), np.float64(-60.0), np.float64(-90.0)) to (Timestamp('2022-12-01 00:00:00'), np.float64(15.0), np.float64(-25.0))
Data columns (total 1 columns):
 #   Column       Dtype  
---  ------       -----  
 0   cloud_cover  float32
dtypes: float32(1)
memory usage: 746.3 MB
None


cloud_cover
time       lat   lon                
1940-01-01 -60.0 -90.00     0.780718
                 -89.75     0.783556
                 -89.50     0.784548
                 -89.25     0.782808
                 -89.00     0.781160

#################### 



,cloud_cover
time,
1940-01-01,78561
1940-02-01,78561
1940-03-01,78561
1940-04-01,78561
1940-05-01,78561
...,...
2022-08-01,78561
2022-09-01,78561
2022-10-01,78561


time
1940-01-01    78561
1940-02-01    78561
1940-03-01    78561
1940-04-01    78561
1940-05-01    78561
              ...  
2022-08-01    78561
2022-09-01    78561
2022-10-01    78561
2022-11-01    78561
2022-12-01    78561
Name: total_registros, Length: 996, dtype: int64

In [7]:
pressao_superficie = xr.open_dataset(f"{path}/treino_surface_pressure.nc").to_dataframe()
geopotencial = xr.open_dataset(f"{path}/treino_geopotential_850.nc").to_dataframe()
umidade = xr.open_dataset(f"{path}/treino_shum_850.nc").to_dataframe()
umidade_relativa = xr.open_dataset(f"{path}/treino_rel_hum_850.nc").to_dataframe()
vento_u = xr.open_dataset(f"{path}/treino_u_850.nc").to_dataframe()
vento_v = xr.open_dataset(f"{path}/treino_v_850.nc").to_dataframe()
temperatura = xr.open_dataset(f"{path}/treino_temperature_850.nc").to_dataframe()
temp_2_metros = xr.open_dataset(f"{path}/treino_t2.nc").to_dataframe()

### Resumo

Rodamos com todas as variáveis e todas seguem o mesmo padrão, então com as de cima simplificamos e só importamos os dados

As conclusões foram que seguem exatamente a descrição da competicão:
1. Existem 996 meses em análise (jan/1940 a dez/2022)
2. Para cada mês, existem latitudes de 60.0°S a 15.0°N, com os valores avançando 0.25 (ex: 60.0,59.75,59.5,59.25,...), então são 60+15=75 valores inteiros com 4 subdivisões para cada, o que resulta em 75*4 = 300, só que +1 contando 60.0°S, então **301 latitudes**
3. Para cada latitudade, existem longitudes de 90°O a 25°O, com os valores avançando 0.25, então são 90-25=65 valores inteiros com 4 subdivisões cada, o que resulta em 65*4 + 1 = **261 longitudes por latitude**
4. Para cada mês, portanto, teremos **301x261 = 78.561 coordenadas por mês**, represenbtando os **78.561 pontos de grade** descritos na competição
5. Para cada mês, em cada um desses **78.561 pontos de grade**, teremos um valor associado, que varia com a tabela (ex: pode ser temperatura, pressão, umidade relativa, etc.)
6. O dataset **`tp`** possui os valores reais dos pontos na grade daquele mês, enquanto **`tp_alvo`** possui um deslocamento de 1 mês, então cada ponto está com o valor daquele mesmo ponto no mês seguinte (ex: o valor de **`tp_alvo`**(jan/1940, lat: 60°S, lon: 90°O) = 3.153801, equivale ao valor de **`tp`**(fev/1940, lat: 60°S, lon: 90°O)=3.153801). Dessa forma, o último mês do dataset, dez/2022, possui os valores reais da grade em **`tp`**, mas todos os valores da grade estão como `NaN` em **`tp_alvo`**, pois não há mês seguinte.
 


### Teste

In [8]:
print("#"*5,"TESTE","#"*5,"\n")

teste = xr.open_dataset(f"{path}/teste_features.nc").to_dataframe()

mostrar_infos_df(teste,"teste")
display(contar_linhas_por_time(teste))
display(contar_valores_por_time(teste))
display(contar_dimensoes_grade(teste))


##### TESTE ##### 



##### TESTE #####
<class 'pandas.core.frame.DataFrame'>
MultiIndex: 1885464 entries, (Timestamp('2023-01-01 00:00:00'), np.float64(-60.0), np.float64(-90.0)) to (Timestamp('2024-12-01 00:00:00'), np.float64(15.0), np.float64(-25.0))
Data columns (total 13 columns):
 #   Column            Dtype         
---  ------            -----         
 0   t2                float32       
 1   cloud_cover       float32       
 2   shum_850          float32       
 3   surface_pressure  float32       
 4   u_850             float32       
 5   v_850             float32       
 6   temperature_850   float32       
 7   rel_hum_850       float32       
 8   geopotential_850  float32       
 9   tp_alvo           float32       
 10  tp_ultima_obs     float32       
 11  time_origem       datetime64[ns]
 12  lag_meses         int16         
dtypes: datetime64[ns](1), float32(11), int16(1)
memory usage: 106.1 MB
None


t2  cloud_cover  shum_850  surface_pressure  \
time       lat   lon                                                           
2023-01-01 -60.0 -90.00  276.325195     0.756066  0.002307         98572.875   
                 -89.75  276.309570     0.754373  0.002310         98577.875   
                 -89.50  276.297852     0.752557  0.002313         98576.875   
                 -89.25  276.286133     0.750558  0.002316         98567.875   
                 -89.00  276.284180     0.749459  0.002321         98563.875   

                             u_850     v_850  temperature_850  rel_hum_850  \
time       lat   lon                                                         
2023-01-01 -60.0 -90.00  10.837509 -1.043722       267.944336    77.139938   
                 -89.75  10.828720 -1.022238       267.959961    77.075485   
                 -89.50  10.815048 -1.002218       267.975586    76.973923   
                 -89.25  10.800400 -0.985128       267.992188    76.891891   
                 -89.00  10.782822 -0.970968       268.010742    76.891891   

                         geopotential_850  tp_alvo  tp_ultima_obs time_origem  \
time       lat   lon                                                            
2023-01-01 -60.0 -90.00      11558.367188      NaN       1.911163  2022-12-01   
                 -89.75      11560.492188      NaN       1.895905  2022-12-01   
                 -89.50      11562.617188      NaN       1.913071  2022-12-01   
                 -89.25      11564.867188      NaN       1.974106  2022-12-01   
                 -89.00      11566.992188      NaN       2.016068  2022-12-01   

                         lag_meses  
time       lat   lon                
2023-01-01 -60.0 -90.00          1  
                 -89.75          1  
                 -89.50          1  
                 -89.25          1  
                 -89.00          1

#################### 



time
2023-01-01    78561
2023-02-01    78561
2023-03-01    78561
2023-04-01    78561
2023-05-01    78561
2023-06-01    78561
2023-07-01    78561
2023-08-01    78561
2023-09-01    78561
2023-10-01    78561
2023-11-01    78561
2023-12-01    78561
2024-01-01    78561
2024-02-01    78561
2024-03-01    78561
2024-04-01    78561
2024-05-01    78561
2024-06-01    78561
2024-07-01    78561
2024-08-01    78561
2024-09-01    78561
2024-10-01    78561
2024-11-01    78561
2024-12-01    78561
Name: total_registros, dtype: int64

,t2,cloud_cover,shum_850,surface_pressure,u_850,v_850,temperature_850,rel_hum_850,geopotential_850,tp_alvo,tp_ultima_obs,time_origem,lag_meses
time,,,,,,,,,,,,,
2023-01-01,78561,78561,78561,78561,78561,78561,78561,78561,78561,0,78561,78561,78561
2023-02-01,78561,78561,78561,78561,78561,78561,78561,78561,78561,0,78561,78561,78561
2023-03-01,78561,78561,78561,78561,78561,78561,78561,78561,78561,0,78561,78561,78561
2023-04-01,78561,78561,78561,78561,78561,78561,78561,78561,78561,0,78561,78561,78561
2023-05-01,78561,78561,78561,78561,78561,78561,78561,78561,78561,0,78561,78561,78561
2023-06-01,78561,78561,78561,78561,78561,78561,78561,78561,78561,0,78561,78561,78561
2023-07-01,78561,78561,78561,78561,78561,78561,78561,78561,78561,0,78561,78561,78561
2023-08-01,78561,78561,78561,78561,78561,78561,78561,78561,78561,0,78561,78561,78561
2023-09-01,78561,78561,78561,78561,78561,78561,78561,78561,78561,0,78561,78561,78561


{'latitudes_por_time': time
 2023-01-01    301
 2023-02-01    301
 2023-03-01    301
 2023-04-01    301
 2023-05-01    301
 2023-06-01    301
 2023-07-01    301
 2023-08-01    301
 2023-09-01    301
 2023-10-01    301
 2023-11-01    301
 2023-12-01    301
 2024-01-01    301
 2024-02-01    301
 2024-03-01    301
 2024-04-01    301
 2024-05-01    301
 2024-06-01    301
 2024-07-01    301
 2024-08-01    301
 2024-09-01    301
 2024-10-01    301
 2024-11-01    301
 2024-12-01    301
 Name: lat, dtype: int64,
 'longitudes_por_latitude': lat
 -60.00    261
 -59.75    261
 -59.50    261
 -59.25    261
 -59.00    261
          ... 
  14.00    261
  14.25    261
  14.50    261
  14.75    261
  15.00    261
 Name: lon, Length: 301, dtype: int64,
 'longitudes_por_time': time
 2023-01-01    78561
 2023-02-01    78561
 2023-03-01    78561
 2023-04-01    78561
 2023-05-01    78561
 2023-06-01    78561
 2023-07-01    78561
 2023-08-01    78561
 2023-09-01    78561
 2023-10-01    78561
 2023-11-01    

### Resumo

Para a tabela de teste, temos uma modificação.

Ao invés de separar em tabelas com cada variável, elas foram unidas em uma única tabela. Dessa forma, cada registro representa um ponto na grade para um determinado mês
 


## ConvLSTM para previsão de precipitação

Cada mês é tratado como um frame `lat x lon`, cada variável climática como um canal e o alvo como a precipitação do mês seguinte.

In [ ]:
"""ConvLSTM para previsão mensal de precipitação na grade lat/lon."""

BASE_FEATURE_FILES = {
    "tp": "treino_tp.nc",
    "cloud_cover": "treino_cloud_cover.nc",
    "surface_pressure": "treino_surface_pressure.nc",
    "geopotential_850": "treino_geopotential_850.nc",
    "shum_850": "treino_shum_850.nc",
    "rel_hum_850": "treino_rel_hum_850.nc",
    "u_850": "treino_u_850.nc",
    "v_850": "treino_v_850.nc",
    "temperature_850": "treino_temperature_850.nc",
    "t2": "treino_t2.nc",
}

INPUT_FEATURE_ORDER = [
    "cloud_cover",
    "geopotential_850",
    "surface_pressure",
    "rel_hum_850",
    "shum_850",
    "u_850",
    "v_850",
    "temperature_850",
    "t2",
]

FEATURE_FILES = BASE_FEATURE_FILES


def selecionar_device():
    """Seleciona TPU, depois GPU e, por último, CPU para o TensorFlow."""
    try:
        resolver = tf.distribute.cluster_resolver.TPUClusterResolver()
        tf.config.experimental_connect_to_cluster(resolver)
        tf.tpu.experimental.initialize_tpu_system(resolver)
        return resolver.master()
    except (ValueError, RuntimeError):
        if tf.config.list_physical_devices("GPU"):
            return "/GPU:0"
        return "/CPU:0"


def abrir_datasets(data_dir: str | Path):
    """Abre os canais de entrada e o alvo sem materializar os dados na memória."""
    data_dir = Path(data_dir)
    features = {}
    for name, filename in FEATURE_FILES.items():
        dataset = xr.open_dataset(data_dir / filename)
        variable = next(iter(dataset.data_vars))
        features[name] = (dataset, variable)

    target = xr.open_dataset(data_dir / "treino_tp_alvo.nc")
    target_variable = str(next(iter(target.data_vars)))
    return features, target, target_variable


def extrair_feature(feature_name: str, feature_datasets, time_indices):
    """Obtém um canal bruto para um intervalo de tempo."""
    if feature_name in INPUT_FEATURE_ORDER:
        dataset, variable = feature_datasets[feature_name]
        return (
            dataset[variable]
            .isel(time=time_indices)
            .transpose("time", "lat", "lon")
            .values
        )

    raise ValueError(f"Canal não suportado: {feature_name}")


class ClimateGridDataset:
    """Monta uma sequência [tempo, lat, lon, canais] sob demanda."""

    def __init__(
        self,
        ends: Iterable[int],
        sequence_length: int,
        feature_datasets,
        target_dataset,
        target_variable: str,
        feature_names: Iterable[str],
        means: dict[str, float],
        stds: dict[str, float],
        target_mean: float,
        target_std: float,
    ):
        self.ends = np.asarray(list(ends), dtype=np.int64)
        self.sequence_length = sequence_length
        self.feature_datasets = feature_datasets
        self.target_dataset = target_dataset
        self.target_variable = target_variable
        self.feature_names = list(feature_names)
        self.means = means
        self.stds = stds
        self.target_mean = target_mean
        self.target_std = target_std

    def __len__(self):
        return len(self.ends)

    def __getitem__(self, index):
        end = int(self.ends[index])
        indices = np.arange(end - self.sequence_length + 1, end + 1)
        channels = []

        for name in self.feature_names:
            frame = np.asarray(extrair_feature(name, self.feature_datasets, indices), dtype=np.float32)
            frame = (frame - self.means[name]) / self.stds[name]
            channels.append(np.nan_to_num(frame, nan=0.0))

        x = np.stack(channels, axis=1)
        y = (
            self.target_dataset[self.target_variable]
            .isel(time=end)
            .transpose("lat", "lon")
            .values
        )
        y = (np.asarray(y, dtype=np.float32) - self.target_mean) / self.target_std
        y = np.nan_to_num(y, nan=0.0)
        x = np.transpose(x, (0, 2, 3, 1))
        return x.astype(np.float32), y[..., np.newaxis].astype(np.float32)


def _criar_tf_dataset(dataset, batch_size, spatial_shape, shuffle=False):
    n_lat, n_lon = spatial_shape
    output_signature = (
        tf.TensorSpec(
            shape=(dataset.sequence_length, n_lat, n_lon, len(dataset.feature_names)),
            dtype=tf.float32,
        ),
        tf.TensorSpec(shape=(n_lat, n_lon, 1), dtype=tf.float32),
    )
    result = tf.data.Dataset.from_generator(
        lambda: (dataset[index] for index in range(len(dataset))),
        output_signature=output_signature,
    )
    if shuffle:
        result = result.shuffle(
            min(max(len(dataset), 1), 128), reshuffle_each_iteration=True
        )
    result = result.batch(batch_size)
    batch_count = (len(dataset) + batch_size - 1) // batch_size
    result = result.apply(tf.data.experimental.assert_cardinality(batch_count))
    return result.prefetch(tf.data.AUTOTUNE)


def preparar_dados(
    data_dir: str | Path,
    sequence_length: int = 3,
    batch_size: int = 2,
    train_fraction: float = 0.8,
    validation_fraction: float = 0.1,
    num_workers: int = 2,
):
    """Cria loaders cronológicos e normaliza usando somente o treino.

    A entrada usa nove canais físicos brutos, sem features derivadas.
    """
    features, target, target_variable = abrir_datasets(data_dir)
    times = pd.DatetimeIndex(features["tp"][0].time.values)
    n_times = len(times)
    train_end = int(n_times * train_fraction)
    validation_end = int(n_times * (train_fraction + validation_fraction))

    valid_ends = np.arange(sequence_length - 1, n_times)
    train_ends = valid_ends[valid_ends < train_end]
    validation_ends = valid_ends[
        (valid_ends >= train_end) & (valid_ends < validation_end)
    ]
    test_ends = valid_ends[valid_ends >= validation_end]

    train_indices = np.arange(train_end)
    means, stds = {}, {}
    for name in INPUT_FEATURE_ORDER:
        values = np.asarray(
            extrair_feature(name, features, train_indices),
            dtype=np.float32,
        )
        means[name] = float(np.nanmean(values))
        stds[name] = max(float(np.nanstd(values)), 1e-6)

    target_values = target[target_variable].isel(time=train_indices).values
    target_mean = float(np.nanmean(target_values))
    target_std = max(float(np.nanstd(target_values)), 1e-6)
    spatial_shape = (
        features["tp"][0].sizes["lat"],
        features["tp"][0].sizes["lon"],
    )
    args = (
        features,
        target,
        target_variable,
        INPUT_FEATURE_ORDER,
        means,
        stds,
        target_mean,
        target_std,
    )

    datasets = {
        "train": ClimateGridDataset(train_ends, sequence_length, *args),
        "validation": ClimateGridDataset(validation_ends, sequence_length, *args),
        "test": ClimateGridDataset(test_ends, sequence_length, *args),
    }
    if not len(datasets["train"]):
        raise ValueError("A divisão de treino não contém amostras.")
    if not len(datasets["validation"]):
        raise ValueError("A divisão de validação não contém amostras.")
    loaders = {
        "train": _criar_tf_dataset(
            datasets["train"], batch_size, spatial_shape, shuffle=True
        ),
        "validation": _criar_tf_dataset(
            datasets["validation"], batch_size, spatial_shape
        ),
        "test": _criar_tf_dataset(datasets["test"], batch_size, spatial_shape),
    }
    metadata = {
        "times": times,
        "n_channels": len(INPUT_FEATURE_ORDER),
        "n_lat": features["tp"][0].sizes["lat"],
        "n_lon": features["tp"][0].sizes["lon"],
        "target_mean": target_mean,
        "target_std": target_std,
        "feature_means": means,
        "feature_stds": stds,
    }
    return loaders, metadata


class ConvLSTMPrecipitacao(tf.keras.Model):
    """ConvLSTM baseada na camada oficial ConvLSTM2D do Keras."""

    def __init__(
        self,
        input_channels: int = 6,
        hidden_channels: int = 32,
        kernel_size: int = 3,
        dropout: float = 0.0,
    ):
        super().__init__()
        self.input_channels = input_channels
        self.hidden_channels = hidden_channels
        self.conv_lstm = tf.keras.layers.ConvLSTM2D(
            filters=hidden_channels,
            kernel_size=(kernel_size, kernel_size),
            padding="same",
            return_sequences=False,
            dropout=dropout,
        )
        self.head = tf.keras.layers.Conv2D(1, kernel_size=1)

    def build(self, input_shape):
        input_shape = tf.TensorShape(input_shape)
        if input_shape.rank != 5:
            raise ValueError(
                "ConvLSTMPrecipitacao espera entradas com forma "
                "(batch, tempo, lat, lon, canais)."
            )
        if any(input_shape[index] is None for index in (2, 3, 4)):
            raise ValueError(
                "As dimensoes tempo, lat, lon e canais precisam ser estaticas "
                "para construir a ConvLSTM."
            )

        self.conv_lstm.build(input_shape)
        conv_output_shape = self.conv_lstm.compute_output_shape(input_shape)
        self.head.build(conv_output_shape)
        super().build(input_shape)

    def call(self, inputs, training=None, mask=None):
        return self.head(self.conv_lstm(inputs, training=training))


def _compilar_modelo(model, learning_rate=1e-3):
    model.compile(
        optimizer=tf.keras.optimizers.Adam(learning_rate=learning_rate),
        loss=tf.keras.losses.MeanSquaredError(),
    )
    return model


def avaliar(model, loader, device, loss_function):
    del device, loss_function
    return float(model.evaluate(loader, verbose=0, return_dict=False))


def treinar(
    model,
    train_loader,
    validation_loader,
    device,
    epochs=10,
    lr=1e-3,
    xla_module=None,
):
    del xla_module
    with tf.device(device):
        if model.optimizer is None:
            _compilar_modelo(model, lr)
        history = model.fit(
            train_loader,
            validation_data=validation_loader,
            epochs=epochs,
            verbose=1,
        ).history
    return {"train": history["loss"], "validation": history["val_loss"]}


def otimizar_hyperparametros(
    train_loader,
    validation_loader,
    input_shape,
    max_trials=8,
    epochs=10,
    diretorio="keras_tuner",
):
    """Busca a configuração ConvLSTM com Keras Tuner e retorna o melhor modelo."""
    try:
        import keras_tuner as kt
    except ImportError as error:
        raise ImportError(
            "Instale Keras Tuner com `pip install keras-tuner`."
        ) from error

    def construir(hp):
        model = ConvLSTMPrecipitacao(
            input_channels=input_shape[-1],
            hidden_channels=hp.Int("hidden_channels", 16, 64, step=16),
            kernel_size=hp.Choice("kernel_size", [3, 5]),
            dropout=hp.Float("dropout", 0.0, 0.3, step=0.1),
        )
        model.build((None, *input_shape))
        return _compilar_modelo(
            model,
            hp.Float("learning_rate", 1e-4, 3e-3, sampling="log"),
        )

    tuner = kt.RandomSearch(
        construir,
        objective="val_loss",
        max_trials=max_trials,
        directory=diretorio,
        project_name="conv_lstm_precipitacao",
        overwrite=True,
    )
    tuner.search(train_loader, validation_data=validation_loader, epochs=epochs, verbose=1)
    melhor_modelo = tuner.get_best_models(num_models=1)[0]
    return melhor_modelo, tuner


def prever(model, loader, device, target_mean, target_std):
    """Gera previsões na escala original do alvo, revertendo a padronização."""
    del device
    predictions = model.predict(loader, verbose=0)
    return predictions * target_std + target_mean


def calcular_rmse(model, loader, device, target_mean, target_std):
    """Calcula RMSE na escala original do alvo para um loader rotulado."""
    soma_erro_quadratico = 0.0
    quantidade_valores = 0
    del device

    for x, y in loader:
        previsao = model(x, training=False).numpy()
        real = y.numpy()
        erro = (previsao - real) * target_std
        soma_erro_quadratico += float(np.sum(erro**2))
        quantidade_valores += erro.size

    if quantidade_valores == 0:
        raise ValueError("O loader não contém amostras para avaliação.")

    return float(np.sqrt(soma_erro_quadratico / quantidade_valores))


def extrair_feature_teste(frame, feature_name: str, latitudes, longitudes):
    """Extrai um canal físico bruto do dataframe de teste."""
    if feature_name in INPUT_FEATURE_ORDER:
        grid = frame[feature_name].unstack("lon").reindex(
            index=latitudes, columns=longitudes
        )
        return grid.to_numpy(dtype=np.float32)

    raise ValueError(f"Canal de teste não suportado: {feature_name}")


def previsoes_teste(
    model,
    teste_df,
    device,
    feature_means,
    feature_stds,
    target_mean,
    target_std,
    batch_size=2,
):
    """Gera previsões para os meses do arquivo oficial de teste."""
    required = {
        "cloud_cover",
        "geopotential_850",
        "surface_pressure",
        "rel_hum_850",
        "shum_850",
        "u_850",
        "v_850",
        "temperature_850",
        "t2",
        "tp_ultima_obs",
    }
    missing = required - set(teste_df.columns)
    if missing:
        raise ValueError(f"Colunas ausentes no teste: {sorted(missing)}")

    times = pd.DatetimeIndex(
        teste_df.index.get_level_values("time").unique()
    ).sort_values()
    times = times[
        (times >= pd.Timestamp("2023-01-01"))
        & (times <= pd.Timestamp("2024-12-01"))
    ]
    latitudes = np.sort(teste_df.index.get_level_values("lat").unique())
    longitudes = np.sort(teste_df.index.get_level_values("lon").unique())
    inputs = []

    for time in times:
        frame = teste_df.loc[time].reset_index().set_index(["lat", "lon"])
        channels = []
        for name in INPUT_FEATURE_ORDER:
            grid = extrair_feature_teste(frame, name, latitudes, longitudes)
            grid = (grid - feature_means[name]) / feature_stds[name]
            channels.append(np.nan_to_num(grid, nan=0.0))

        grid = np.stack(channels, axis=-1)
        inputs.append(grid[np.newaxis, ...])

    del device
    predicted_grids = []
    for start in range(0, len(inputs), batch_size):
        x = np.concatenate(inputs[start : start + batch_size], axis=0)
        predicted_grids.append(
            model.predict(x, verbose=0)[:, :, :, 0] * target_std + target_mean
        )

    return np.concatenate(predicted_grids)


def avaliar_teste_contra_ultima_obs(
    model,
    teste_df,
    device,
    feature_means,
    feature_stds,
    target_mean,
    target_std,
    batch_size=2,
):
    """Prevê tp no teste e compara com `tp_ultima_obs` por mês.

    Esta é uma comparação de referência/persistência. `tp_ultima_obs` é o
    último valor observado, não o valor real do mês previsto.
    """
    required = {
        "cloud_cover",
        "geopotential_850",
        "surface_pressure",
        "rel_hum_850",
        "shum_850",
        "u_850",
        "v_850",
        "temperature_850",
        "t2",
        "tp_ultima_obs",
    }
    missing = required - set(teste_df.columns)
    if missing:
        raise ValueError(f"Colunas ausentes no teste: {sorted(missing)}")

    times = pd.DatetimeIndex(teste_df.index.get_level_values("time").unique()).sort_values()
    times = times[
        (times >= pd.Timestamp("2023-01-01"))
        & (times <= pd.Timestamp("2024-12-01"))
    ]
    latitudes = np.sort(teste_df.index.get_level_values("lat").unique())
    longitudes = np.sort(teste_df.index.get_level_values("lon").unique())
    predictions, references, dates = [], [], []

    for time in times:
        frame = teste_df.loc[time].reset_index().set_index(["lat", "lon"])
        channels = []
        for name in INPUT_FEATURE_ORDER:
            grid = extrair_feature_teste(frame, name, latitudes, longitudes)
            grid = (grid - feature_means[name]) / feature_stds[name]
            channels.append(np.nan_to_num(grid, nan=0.0))

        grid = np.stack(channels, axis=-1)
        x = grid[np.newaxis, np.newaxis, ...]
        reference = frame["tp_ultima_obs"].unstack("lon").reindex(
            index=latitudes, columns=longitudes
        ).to_numpy(dtype=np.float32)
        predictions.append(x)
        references.append(reference)
        dates.append(time)

    del device
    predicted_grids = []
    for start in range(0, len(predictions), batch_size):
        x = np.concatenate(predictions[start : start + batch_size], axis=0)
        predicted_grids.append(
            model.predict(x, verbose=0)[:, :, :, 0] * target_std + target_mean
        )

    predicted_grids = np.concatenate(predicted_grids)
    references = np.asarray(references)
    errors = predicted_grids - references
    monthly_rmse = np.sqrt(np.mean(errors**2, axis=(1, 2)))
    result = pd.DataFrame(
        {"time": dates, "rmse_tp_vs_tp_ultima_obs": monthly_rmse}
    ).set_index("time")
    global_rmse = float(np.sqrt(np.mean(errors**2)))
    return global_rmse, result, predicted_grids


def plotar_historico(history):
    """Plota a perda de treino e validação ao longo das épocas."""
    import matplotlib.pyplot as plt

    epocas = np.arange(1, len(history["train"]) + 1)
    plt.figure(figsize=(9, 5))
    plt.plot(epocas, history["train"], marker="o", label="Treinamento")
    plt.plot(epocas, history["validation"], marker="o", label="Validação")
    plt.xlabel("Época")
    plt.ylabel("MSE normalizado")
    plt.title("Perda durante o treinamento")
    plt.grid(alpha=0.3)
    plt.legend()
    plt.tight_layout()
    plt.show()


def diagnosticar_device(model, device):
    """Confirma se modelo, CUDA e um batch estão no dispositivo esperado."""
    print("device selecionado:", device)
    print("dispositivos TensorFlow:", tf.config.list_logical_devices())
    print("modelo compilado:", model.optimizer is not None)


In [10]:
device = selecionar_device()
# O mês t prevê o mês t+1; por isso cada amostra usa apenas um mês de entrada.
loaders, convlstm_metadata = preparar_dados(path, sequence_length=1, batch_size=2, num_workers=2)
input_shape = (
    1,
    convlstm_metadata["n_lat"],
    convlstm_metadata["n_lon"],
    convlstm_metadata["n_channels"],
)
model, tuner = otimizar_hyperparametros(
    loaders["train"],
    loaders["validation"],
    input_shape=input_shape,
    max_trials=8,
    epochs=10,
)
print("Melhores hiperparâmetros:", tuner.get_best_hyperparameters(1)[0].values)
print(convlstm_metadata)
diagnosticar_device(model, device)

Trial 3 Complete [03h 51m 54s]
val_loss: 0.46605992317199707

Best val_loss So Far: 0.42012420296669006
Total elapsed time: 11h 34m 38s

Search: Running Trial #4

Value             |Best Value So Far |Hyperparameter
64                |64                |hidden_channels
3                 |5                 |kernel_size
0                 |0.1               |dropout
0.0003773         |0.0010512         |learning_rate



Epoch 1/10



  1/398 ━━━━━━━━━━━━━━━━━━━━ 22:44:24 206s/step - loss: 1.1212


  2/398 ━━━━━━━━━━━━━━━━━━━━ 1:57 297ms/step - loss: 1.1034   


  3/398 ━━━━━━━━━━━━━━━━━━━━ 11:02 2s/step - loss: 1.0933  


  4/398 ━━━━━━━━━━━━━━━━━━━━ 14:14 2s/step - loss: 1.0765


  5/398 ━━━━━━━━━━━━━━━━━━━━ 15:44 2s/step - loss: 1.0610


  6/398 ━━━━━━━━━━━━━━━━━━━━ 16:36 3s/step - loss: 1.0479


  7/398 ━━━━━━━━━━━━━━━━━━━━ 17:12 3s/step - loss: 1.0426


  8/398 ━━━━━━━━━━━━━━━━━━━━ 17:33 3s/step - loss: 1.0378


  9/398 ━━━━━━━━━━━━━━━━━━━━ 17:50 3s/step - loss: 1.0327


 10/398 ━━━━━━━━━━━━━━━━━━━━ 18:01 3s/step - loss: 1.0286


 11/398 ━━━━━━━━━━━━━━━━━━━━ 18:06 3s/step - loss: 1.0229


 12/398 ━━━━━━━━━━━━━━━━━━━━ 18:11 3s/step - loss: 1.0155


 13/398 ━━━━━━━━━━━━━━━━━━━━ 18:16 3s/step - loss: 1.0104


 14/398 ━━━━━━━━━━━━━━━━━━━━ 18:18 3s/step - loss: 1.0052


 15/398 ━━━━━━━━━━━━━━━━━━━━ 18:20 3s/step - loss: 1.0004


 16/398 ━━━━━━━━━━━━━━━━━━━━ 18:23 3s/step - loss: 0.9960


 17/398 ━━━━━━━━━━━━━━━━━━━━ 18:26 3s/step - loss: 0.9911


 18/398 ━━━━━━━━━━━━━━━━━━━━ 18:28 3s/step - loss: 0.9860


 19/398 ━━━━━━━━━━━━━━━━━━━━ 18:29 3s/step - loss: 0.9806


 20/398 ━━━━━━━━━━━━━━━━━━━━ 18:29 3s/step - loss: 0.9752


 21/398 ━━━━━━━━━━━━━━━━━━━━ 18:29 3s/step - loss: 0.9698


 22/398 ━━━━━━━━━━━━━━━━━━━━ 18:29 3s/step - loss: 0.9645


 23/398 ━━━━━━━━━━━━━━━━━━━━ 18:29 3s/step - loss: 0.9589


 24/398 ━━━━━━━━━━━━━━━━━━━━ 18:27 3s/step - loss: 0.9535


 25/398 ━━━━━━━━━━━━━━━━━━━━ 18:25 3s/step - loss: 0.9482


 26/398 ━━━━━━━━━━━━━━━━━━━━ 18:24 3s/step - loss: 0.9429


 27/398 ━━━━━━━━━━━━━━━━━━━━ 18:22 3s/step - loss: 0.9375


 28/398 ━━━━━━━━━━━━━━━━━━━━ 18:19 3s/step - loss: 0.9324


 29/398 ━━━━━━━━━━━━━━━━━━━━ 18:17 3s/step - loss: 0.9276


 30/398 ━━━━━━━━━━━━━━━━━━━━ 18:15 3s/step - loss: 0.9231


 31/398 ━━━━━━━━━━━━━━━━━━━━ 18:12 3s/step - loss: 0.9189


 32/398 ━━━━━━━━━━━━━━━━━━━━ 18:09 3s/step - loss: 0.9147


 33/398 ━━━━━━━━━━━━━━━━━━━━ 18:07 3s/step - loss: 0.9109


 34/398 ━━━━━━━━━━━━━━━━━━━━ 18:04 3s/step - loss: 0.9071


 35/398 ━━━━━━━━━━━━━━━━━━━━ 18:01 3s/step - loss: 0.9034


 36/398 ━━━━━━━━━━━━━━━━━━━━ 17:59 3s/step - loss: 0.8999


 37/398 ━━━━━━━━━━━━━━━━━━━━ 17:57 3s/step - loss: 0.8963


 38/398 ━━━━━━━━━━━━━━━━━━━━ 17:54 3s/step - loss: 0.8929


 39/398 ━━━━━━━━━━━━━━━━━━━━ 17:52 3s/step - loss: 0.8896


 40/398 ━━━━━━━━━━━━━━━━━━━━ 17:50 3s/step - loss: 0.8863


 41/398 ━━━━━━━━━━━━━━━━━━━━ 17:48 3s/step - loss: 0.8832


 42/398 ━━━━━━━━━━━━━━━━━━━━ 17:46 3s/step - loss: 0.8800


 43/398 ━━━━━━━━━━━━━━━━━━━━ 17:44 3s/step - loss: 0.8770


 44/398 ━━━━━━━━━━━━━━━━━━━━ 17:41 3s/step - loss: 0.8740


 45/398 ━━━━━━━━━━━━━━━━━━━━ 17:39 3s/step - loss: 0.8710


 46/398 ━━━━━━━━━━━━━━━━━━━━ 17:37 3s/step - loss: 0.8681


 47/398 ━━━━━━━━━━━━━━━━━━━━ 17:35 3s/step - loss: 0.8654


 48/398 ━━━━━━━━━━━━━━━━━━━━ 17:33 3s/step - loss: 0.8627


 49/398 ━━━━━━━━━━━━━━━━━━━━ 17:31 3s/step - loss: 0.8600


 50/398 ━━━━━━━━━━━━━━━━━━━━ 17:28 3s/step - loss: 0.8574


 51/398 ━━━━━━━━━━━━━━━━━━━━ 17:25 3s/step - loss: 0.8549


 52/398 ━━━━━━━━━━━━━━━━━━━━ 17:23 3s/step - loss: 0.8524


 53/398 ━━━━━━━━━━━━━━━━━━━━ 17:20 3s/step - loss: 0.8500


 54/398 ━━━━━━━━━━━━━━━━━━━━ 17:17 3s/step - loss: 0.8476


 55/398 ━━━━━━━━━━━━━━━━━━━━ 17:14 3s/step - loss: 0.8453


 56/398 ━━━━━━━━━━━━━━━━━━━━ 17:11 3s/step - loss: 0.8430


 57/398 ━━━━━━━━━━━━━━━━━━━━ 17:09 3s/step - loss: 0.8407


 58/398 ━━━━━━━━━━━━━━━━━━━━ 17:06 3s/step - loss: 0.8386


 59/398 ━━━━━━━━━━━━━━━━━━━━ 17:04 3s/step - loss: 0.8366


 60/398 ━━━━━━━━━━━━━━━━━━━━ 17:01 3s/step - loss: 0.8346


 61/398 ━━━━━━━━━━━━━━━━━━━━ 16:58 3s/step - loss: 0.8326


 62/398 ━━━━━━━━━━━━━━━━━━━━ 16:56 3s/step - loss: 0.8306


 63/398 ━━━━━━━━━━━━━━━━━━━━ 16:52 3s/step - loss: 0.8287


 64/398 ━━━━━━━━━━━━━━━━━━━━ 16:50 3s/step - loss: 0.8267


 65/398 ━━━━━━━━━━━━━━━━━━━━ 16:47 3s/step - loss: 0.8248


 66/398 ━━━━━━━━━━━━━━━━━━━━ 16:44 3s/step - loss: 0.8230


 67/398 ━━━━━━━━━━━━━━━━━━━━ 16:41 3s/step - loss: 0.8211


 68/398 ━━━━━━━━━━━━━━━━━━━━ 16:39 3s/step - loss: 0.8192


 69/398 ━━━━━━━━━━━━━━━━━━━━ 16:36 3s/step - loss: 0.8174


 70/398 ━━━━━━━━━━━━━━━━━━━━ 16:34 3s/step - loss: 0.8157


 71/398 ━━━━━━━━━━━━━━━━━━━━ 16:31 3s/step - loss: 0.8139


 72/398 ━━━━━━━━━━━━━━━━━━━━ 16:28 3s/step - loss: 0.8121


 73/398 ━━━━━━━━━━━━━━━━━━━━ 16:26 3s/step - loss: 0.8104


 74/398 ━━━━━━━━━━━━━━━━━━━━ 16:23 3s/step - loss: 0.8087


 75/398 ━━━━━━━━━━━━━━━━━━━━ 16:20 3s/step - loss: 0.8071


 76/398 ━━━━━━━━━━━━━━━━━━━━ 16:17 3s/step - loss: 0.8054


 77/398 ━━━━━━━━━━━━━━━━━━━━ 16:14 3s/step - loss: 0.8038


 78/398 ━━━━━━━━━━━━━━━━━━━━ 16:12 3s/step - loss: 0.8022


 79/398 ━━━━━━━━━━━━━━━━━━━━ 16:09 3s/step - loss: 0.8005


 80/398 ━━━━━━━━━━━━━━━━━━━━ 16:07 3s/step - loss: 0.7989


 81/398 ━━━━━━━━━━━━━━━━━━━━ 16:04 3s/step - loss: 0.7973


 82/398 ━━━━━━━━━━━━━━━━━━━━ 16:02 3s/step - loss: 0.7957


 83/398 ━━━━━━━━━━━━━━━━━━━━ 15:59 3s/step - loss: 0.7941


 84/398 ━━━━━━━━━━━━━━━━━━━━ 15:56 3s/step - loss: 0.7926


 85/398 ━━━━━━━━━━━━━━━━━━━━ 15:53 3s/step - loss: 0.7911


 86/398 ━━━━━━━━━━━━━━━━━━━━ 15:50 3s/step - loss: 0.7896


 87/398 ━━━━━━━━━━━━━━━━━━━━ 15:47 3s/step - loss: 0.7881


 88/398 ━━━━━━━━━━━━━━━━━━━━ 15:44 3s/step - loss: 0.7867


 89/398 ━━━━━━━━━━━━━━━━━━━━ 15:41 3s/step - loss: 0.7853


 90/398 ━━━━━━━━━━━━━━━━━━━━ 15:39 3s/step - loss: 0.7839


 91/398 ━━━━━━━━━━━━━━━━━━━━ 15:36 3s/step - loss: 0.7826


 92/398 ━━━━━━━━━━━━━━━━━━━━ 15:33 3s/step - loss: 0.7813


 93/398 ━━━━━━━━━━━━━━━━━━━━ 15:30 3s/step - loss: 0.7800


 94/398 ━━━━━━━━━━━━━━━━━━━━ 15:27 3s/step - loss: 0.7788


 95/398 ━━━━━━━━━━━━━━━━━━━━ 15:24 3s/step - loss: 0.7775


 96/398 ━━━━━━━━━━━━━━━━━━━━ 15:21 3s/step - loss: 0.7763


 97/398 ━━━━━━━━━━━━━━━━━━━━ 15:19 3s/step - loss: 0.7751


 98/398 ━━━━━━━━━━━━━━━━━━━━ 15:16 3s/step - loss: 0.7739


 99/398 ━━━━━━━━━━━━━━━━━━━━ 15:13 3s/step - loss: 0.7727


100/398 ━━━━━━━━━━━━━━━━━━━━ 15:10 3s/step - loss: 0.7716


101/398 ━━━━━━━━━━━━━━━━━━━━ 15:07 3s/step - loss: 0.7704


102/398 ━━━━━━━━━━━━━━━━━━━━ 15:04 3s/step - loss: 0.7693


103/398 ━━━━━━━━━━━━━━━━━━━━ 15:01 3s/step - loss: 0.7682


104/398 ━━━━━━━━━━━━━━━━━━━━ 14:58 3s/step - loss: 0.7671


105/398 ━━━━━━━━━━━━━━━━━━━━ 14:55 3s/step - loss: 0.7660


106/398 ━━━━━━━━━━━━━━━━━━━━ 14:52 3s/step - loss: 0.7649


107/398 ━━━━━━━━━━━━━━━━━━━━ 14:49 3s/step - loss: 0.7638


108/398 ━━━━━━━━━━━━━━━━━━━━ 14:46 3s/step - loss: 0.7627


109/398 ━━━━━━━━━━━━━━━━━━━━ 14:43 3s/step - loss: 0.7617


110/398 ━━━━━━━━━━━━━━━━━━━━ 14:40 3s/step - loss: 0.7606


111/398 ━━━━━━━━━━━━━━━━━━━━ 14:37 3s/step - loss: 0.7596


112/398 ━━━━━━━━━━━━━━━━━━━━ 14:34 3s/step - loss: 0.7585


113/398 ━━━━━━━━━━━━━━━━━━━━ 14:31 3s/step - loss: 0.7575


114/398 ━━━━━━━━━━━━━━━━━━━━ 14:28 3s/step - loss: 0.7565


115/398 ━━━━━━━━━━━━━━━━━━━━ 14:26 3s/step - loss: 0.7554


116/398 ━━━━━━━━━━━━━━━━━━━━ 14:23 3s/step - loss: 0.7544


117/398 ━━━━━━━━━━━━━━━━━━━━ 14:20 3s/step - loss: 0.7534


118/398 ━━━━━━━━━━━━━━━━━━━━ 14:17 3s/step - loss: 0.7525


119/398 ━━━━━━━━━━━━━━━━━━━━ 14:14 3s/step - loss: 0.7515


120/398 ━━━━━━━━━━━━━━━━━━━━ 14:11 3s/step - loss: 0.7505


121/398 ━━━━━━━━━━━━━━━━━━━━ 14:08 3s/step - loss: 0.7496


122/398 ━━━━━━━━━━━━━━━━━━━━ 14:05 3s/step - loss: 0.7487


123/398 ━━━━━━━━━━━━━━━━━━━━ 14:03 3s/step - loss: 0.7477


124/398 ━━━━━━━━━━━━━━━━━━━━ 14:00 3s/step - loss: 0.7468


125/398 ━━━━━━━━━━━━━━━━━━━━ 13:57 3s/step - loss: 0.7459


126/398 ━━━━━━━━━━━━━━━━━━━━ 13:54 3s/step - loss: 0.7450


127/398 ━━━━━━━━━━━━━━━━━━━━ 13:51 3s/step - loss: 0.7441


128/398 ━━━━━━━━━━━━━━━━━━━━ 13:48 3s/step - loss: 0.7432


129/398 ━━━━━━━━━━━━━━━━━━━━ 13:45 3s/step - loss: 0.7423


130/398 ━━━━━━━━━━━━━━━━━━━━ 13:42 3s/step - loss: 0.7414


131/398 ━━━━━━━━━━━━━━━━━━━━ 13:39 3s/step - loss: 0.7406


132/398 ━━━━━━━━━━━━━━━━━━━━ 13:36 3s/step - loss: 0.7397


133/398 ━━━━━━━━━━━━━━━━━━━━ 13:33 3s/step - loss: 0.7389


134/398 ━━━━━━━━━━━━━━━━━━━━ 13:30 3s/step - loss: 0.7381


135/398 ━━━━━━━━━━━━━━━━━━━━ 13:27 3s/step - loss: 0.7372


136/398 ━━━━━━━━━━━━━━━━━━━━ 13:24 3s/step - loss: 0.7364


137/398 ━━━━━━━━━━━━━━━━━━━━ 13:21 3s/step - loss: 0.7356


138/398 ━━━━━━━━━━━━━━━━━━━━ 13:18 3s/step - loss: 0.7347


139/398 ━━━━━━━━━━━━━━━━━━━━ 13:15 3s/step - loss: 0.7339


140/398 ━━━━━━━━━━━━━━━━━━━━ 13:12 3s/step - loss: 0.7331


141/398 ━━━━━━━━━━━━━━━━━━━━ 13:09 3s/step - loss: 0.7324


142/398 ━━━━━━━━━━━━━━━━━━━━ 13:05 3s/step - loss: 0.7316


143/398 ━━━━━━━━━━━━━━━━━━━━ 13:02 3s/step - loss: 0.7308


144/398 ━━━━━━━━━━━━━━━━━━━━ 12:59 3s/step - loss: 0.7301


145/398 ━━━━━━━━━━━━━━━━━━━━ 12:56 3s/step - loss: 0.7293


146/398 ━━━━━━━━━━━━━━━━━━━━ 12:53 3s/step - loss: 0.7285


147/398 ━━━━━━━━━━━━━━━━━━━━ 12:50 3s/step - loss: 0.7278


148/398 ━━━━━━━━━━━━━━━━━━━━ 12:47 3s/step - loss: 0.7270


149/398 ━━━━━━━━━━━━━━━━━━━━ 12:44 3s/step - loss: 0.7263


150/398 ━━━━━━━━━━━━━━━━━━━━ 12:42 3s/step - loss: 0.7256


151/398 ━━━━━━━━━━━━━━━━━━━━ 12:38 3s/step - loss: 0.7249


152/398 ━━━━━━━━━━━━━━━━━━━━ 12:36 3s/step - loss: 0.7241


153/398 ━━━━━━━━━━━━━━━━━━━━ 12:32 3s/step - loss: 0.7235


154/398 ━━━━━━━━━━━━━━━━━━━━ 12:30 3s/step - loss: 0.7228


155/398 ━━━━━━━━━━━━━━━━━━━━ 12:27 3s/step - loss: 0.7221


156/398 ━━━━━━━━━━━━━━━━━━━━ 12:24 3s/step - loss: 0.7214


157/398 ━━━━━━━━━━━━━━━━━━━━ 12:20 3s/step - loss: 0.7208


158/398 ━━━━━━━━━━━━━━━━━━━━ 12:17 3s/step - loss: 0.7201


159/398 ━━━━━━━━━━━━━━━━━━━━ 12:14 3s/step - loss: 0.7194


160/398 ━━━━━━━━━━━━━━━━━━━━ 12:11 3s/step - loss: 0.7188


161/398 ━━━━━━━━━━━━━━━━━━━━ 12:08 3s/step - loss: 0.7181


162/398 ━━━━━━━━━━━━━━━━━━━━ 12:05 3s/step - loss: 0.7175


163/398 ━━━━━━━━━━━━━━━━━━━━ 12:02 3s/step - loss: 0.7169


164/398 ━━━━━━━━━━━━━━━━━━━━ 11:59 3s/step - loss: 0.7162


165/398 ━━━━━━━━━━━━━━━━━━━━ 11:56 3s/step - loss: 0.7156


166/398 ━━━━━━━━━━━━━━━━━━━━ 11:53 3s/step - loss: 0.7150


167/398 ━━━━━━━━━━━━━━━━━━━━ 11:50 3s/step - loss: 0.7144


168/398 ━━━━━━━━━━━━━━━━━━━━ 11:47 3s/step - loss: 0.7138


169/398 ━━━━━━━━━━━━━━━━━━━━ 11:44 3s/step - loss: 0.7132


170/398 ━━━━━━━━━━━━━━━━━━━━ 11:40 3s/step - loss: 0.7126


171/398 ━━━━━━━━━━━━━━━━━━━━ 11:37 3s/step - loss: 0.7120


172/398 ━━━━━━━━━━━━━━━━━━━━ 11:34 3s/step - loss: 0.7114


173/398 ━━━━━━━━━━━━━━━━━━━━ 11:31 3s/step - loss: 0.7108


174/398 ━━━━━━━━━━━━━━━━━━━━ 11:28 3s/step - loss: 0.7103


175/398 ━━━━━━━━━━━━━━━━━━━━ 11:25 3s/step - loss: 0.7097


176/398 ━━━━━━━━━━━━━━━━━━━━ 11:22 3s/step - loss: 0.7091


177/398 ━━━━━━━━━━━━━━━━━━━━ 11:18 3s/step - loss: 0.7086


178/398 ━━━━━━━━━━━━━━━━━━━━ 11:15 3s/step - loss: 0.7080


179/398 ━━━━━━━━━━━━━━━━━━━━ 11:12 3s/step - loss: 0.7075


180/398 ━━━━━━━━━━━━━━━━━━━━ 11:09 3s/step - loss: 0.7069


181/398 ━━━━━━━━━━━━━━━━━━━━ 11:06 3s/step - loss: 0.7064


182/398 ━━━━━━━━━━━━━━━━━━━━ 11:03 3s/step - loss: 0.7058


183/398 ━━━━━━━━━━━━━━━━━━━━ 11:00 3s/step - loss: 0.7053


184/398 ━━━━━━━━━━━━━━━━━━━━ 10:57 3s/step - loss: 0.7048


185/398 ━━━━━━━━━━━━━━━━━━━━ 10:54 3s/step - loss: 0.7043


186/398 ━━━━━━━━━━━━━━━━━━━━ 10:51 3s/step - loss: 0.7037


187/398 ━━━━━━━━━━━━━━━━━━━━ 10:48 3s/step - loss: 0.7032


188/398 ━━━━━━━━━━━━━━━━━━━━ 10:45 3s/step - loss: 0.7027


189/398 ━━━━━━━━━━━━━━━━━━━━ 10:42 3s/step - loss: 0.7022


190/398 ━━━━━━━━━━━━━━━━━━━━ 10:39 3s/step - loss: 0.7017


191/398 ━━━━━━━━━━━━━━━━━━━━ 10:36 3s/step - loss: 0.7011


192/398 ━━━━━━━━━━━━━━━━━━━━ 10:33 3s/step - loss: 0.7006


193/398 ━━━━━━━━━━━━━━━━━━━━ 10:30 3s/step - loss: 0.7001


194/398 ━━━━━━━━━━━━━━━━━━━━ 10:27 3s/step - loss: 0.6996


195/398 ━━━━━━━━━━━━━━━━━━━━ 10:23 3s/step - loss: 0.6991


196/398 ━━━━━━━━━━━━━━━━━━━━ 10:20 3s/step - loss: 0.6986


197/398 ━━━━━━━━━━━━━━━━━━━━ 10:17 3s/step - loss: 0.6981


198/398 ━━━━━━━━━━━━━━━━━━━━ 10:14 3s/step - loss: 0.6976


199/398 ━━━━━━━━━━━━━━━━━━━━ 10:11 3s/step - loss: 0.6971


200/398 ━━━━━━━━━━━━━━━━━━━━ 10:08 3s/step - loss: 0.6966


201/398 ━━━━━━━━━━━━━━━━━━━━ 10:05 3s/step - loss: 0.6961


202/398 ━━━━━━━━━━━━━━━━━━━━ 10:02 3s/step - loss: 0.6957


203/398 ━━━━━━━━━━━━━━━━━━━━ 9:59 3s/step - loss: 0.6952 


204/398 ━━━━━━━━━━━━━━━━━━━━ 9:56 3s/step - loss: 0.6947


205/398 ━━━━━━━━━━━━━━━━━━━━ 9:53 3s/step - loss: 0.6942


206/398 ━━━━━━━━━━━━━━━━━━━━ 9:50 3s/step - loss: 0.6937


207/398 ━━━━━━━━━━━━━━━━━━━━ 9:47 3s/step - loss: 0.6933


208/398 ━━━━━━━━━━━━━━━━━━━━ 9:44 3s/step - loss: 0.6928


209/398 ━━━━━━━━━━━━━━━━━━━━ 9:41 3s/step - loss: 0.6923


210/398 ━━━━━━━━━━━━━━━━━━━━ 9:38 3s/step - loss: 0.6919


211/398 ━━━━━━━━━━━━━━━━━━━━ 9:35 3s/step - loss: 0.6914


212/398 ━━━━━━━━━━━━━━━━━━━━ 9:32 3s/step - loss: 0.6909


213/398 ━━━━━━━━━━━━━━━━━━━━ 9:29 3s/step - loss: 0.6905


214/398 ━━━━━━━━━━━━━━━━━━━━ 9:26 3s/step - loss: 0.6900


215/398 ━━━━━━━━━━━━━━━━━━━━ 9:23 3s/step - loss: 0.6896


216/398 ━━━━━━━━━━━━━━━━━━━━ 9:20 3s/step - loss: 0.6891


217/398 ━━━━━━━━━━━━━━━━━━━━ 9:17 3s/step - loss: 0.6887


218/398 ━━━━━━━━━━━━━━━━━━━━ 9:14 3s/step - loss: 0.6882


219/398 ━━━━━━━━━━━━━━━━━━━━ 9:11 3s/step - loss: 0.6878


220/398 ━━━━━━━━━━━━━━━━━━━━ 9:08 3s/step - loss: 0.6874


221/398 ━━━━━━━━━━━━━━━━━━━━ 9:04 3s/step - loss: 0.6869


222/398 ━━━━━━━━━━━━━━━━━━━━ 9:01 3s/step - loss: 0.6865


223/398 ━━━━━━━━━━━━━━━━━━━━ 8:58 3s/step - loss: 0.6861


224/398 ━━━━━━━━━━━━━━━━━━━━ 8:55 3s/step - loss: 0.6856


225/398 ━━━━━━━━━━━━━━━━━━━━ 8:52 3s/step - loss: 0.6852


226/398 ━━━━━━━━━━━━━━━━━━━━ 8:49 3s/step - loss: 0.6848


227/398 ━━━━━━━━━━━━━━━━━━━━ 8:46 3s/step - loss: 0.6844


228/398 ━━━━━━━━━━━━━━━━━━━━ 8:43 3s/step - loss: 0.6840


229/398 ━━━━━━━━━━━━━━━━━━━━ 8:40 3s/step - loss: 0.6836


230/398 ━━━━━━━━━━━━━━━━━━━━ 8:37 3s/step - loss: 0.6831


231/398 ━━━━━━━━━━━━━━━━━━━━ 8:34 3s/step - loss: 0.6827


232/398 ━━━━━━━━━━━━━━━━━━━━ 8:31 3s/step - loss: 0.6823


233/398 ━━━━━━━━━━━━━━━━━━━━ 8:28 3s/step - loss: 0.6819


234/398 ━━━━━━━━━━━━━━━━━━━━ 8:25 3s/step - loss: 0.6815


235/398 ━━━━━━━━━━━━━━━━━━━━ 8:21 3s/step - loss: 0.6811


236/398 ━━━━━━━━━━━━━━━━━━━━ 8:18 3s/step - loss: 0.6807


237/398 ━━━━━━━━━━━━━━━━━━━━ 8:15 3s/step - loss: 0.6803


238/398 ━━━━━━━━━━━━━━━━━━━━ 8:12 3s/step - loss: 0.6799


239/398 ━━━━━━━━━━━━━━━━━━━━ 8:09 3s/step - loss: 0.6795


240/398 ━━━━━━━━━━━━━━━━━━━━ 8:06 3s/step - loss: 0.6791


241/398 ━━━━━━━━━━━━━━━━━━━━ 8:03 3s/step - loss: 0.6787


242/398 ━━━━━━━━━━━━━━━━━━━━ 8:00 3s/step - loss: 0.6784

### OBSERVAÇÃO

O kaggle não rodou as 8 iterações, mas apenas 4. 

Porém, para envio do `submission.csv` coletamos os arquivos `trial.json` e `computed.weights.h5` gerados pelo Keras Turner, fizemos um ajuste temporário nas células abaixo para o modelo ser carregado com esses parâmetros e rodamos até o final. 

Sendo assim, considere as células abaixo como o fluxo previsto, mas o motivo de não ter output explícito se deve pelo que foi posto acima.

In [ ]:
melhor_trial = tuner.oracle.get_best_trials(num_trials=1)[0]

historico = {
    "train": [
        observacao.value[0]
        for observacao in melhor_trial.metrics.get_history("loss")
    ],
    "validation": [
        observacao.value[0]
        for observacao in melhor_trial.metrics.get_history("val_loss")
    ],
}

plotar_historico(historico)

In [ ]:
# Previsões para os meses do conjunto de teste, após o treinamento.
previsoes = prever(
    model, loaders["test"], device,
    convlstm_metadata["target_mean"],
    convlstm_metadata["target_std"],
)
print(previsoes.shape)  # [amostras, 1, latitude, longitude]


### RMSE no conjunto de teste interno (10% de 1940 a 2022)

Este `test_loader` é a parte final do período de treino, que possui `tp_alvo` (10% do conjunto dedicado a teste)

In [ ]:
rmse_teste = calcular_rmse(
    model,
    loaders["test"],
    device,
    convlstm_metadata["target_mean"],
    convlstm_metadata["target_std"],
)
print(f"RMSE no teste: {rmse_teste:.6f} mm/dia")

## RMSE no conjunto oficial de teste (2023/2024)

Esta comparação usa `tp_ultima_obs` como referência de persistência da base de teste.

In [ ]:
rmse_tp_ultima_obs, rmse_mensal, previsoes_teste = avaliar_teste_contra_ultima_obs(
    model,
    teste,
    device,
    convlstm_metadata["feature_means"],
    convlstm_metadata["feature_stds"],
    convlstm_metadata["target_mean"],
    convlstm_metadata["target_std"],
)
print(f"RMSE global de tp contra tp_ultima_obs: {rmse_tp_ultima_obs:.6f} mm/dia")
display(rmse_mensal)
print("Formato das previsões:", previsoes_teste.shape)


### Gerando o arquivo de submissão para 2023–2024

As previsões são gravadas no mesmo formato de `sample_submission.csv`: uma linha por mês e ponto da grade.

In [ ]:
df_submission = pd.read_csv(f"{path}/sample_submission.csv")

datas_submissao = pd.DatetimeIndex(
    teste.index.get_level_values("time").unique()
).sort_values()
datas_submissao = datas_submissao[
    (datas_submissao >= pd.Timestamp("2023-01-01"))
    & (datas_submissao <= pd.Timestamp("2024-12-01"))
]
latitudes_submissao = np.sort(teste.index.get_level_values("lat").unique())
longitudes_submissao = np.sort(teste.index.get_level_values("lon").unique())

previsoes_teste = np.asarray(previsoes_teste)
esperado = (
    len(datas_submissao),
    len(latitudes_submissao),
    len(longitudes_submissao),
)
if previsoes_teste.shape != esperado:
    raise ValueError(
        f"Formato inesperado das previsões: {previsoes_teste.shape}; "
        f"esperado: {esperado}"
    )

ids_esperados = [
    f"{data:%Y_%m}_{latitude:.2f}_{longitude:.2f}"
    for data in datas_submissao
    for latitude in latitudes_submissao
    for longitude in longitudes_submissao
]
if ids_esperados != df_submission["id"].tolist():
    raise ValueError("A ordem das coordenadas não coincide com sample_submission.csv")

# O flatten preserva a ordem mês, latitude e longitude usada nos IDs.
df_submission_convlstm = df_submission.copy()
df_submission_convlstm["tp_mm_day"] = previsoes_teste.reshape(-1)

if len(df_submission_convlstm) != 1_885_464:
    raise ValueError(
        f"Quantidade inesperada de previsões: {len(df_submission_convlstm)}"
    )

arquivo_submissao = "submission.csv"
df_submission_convlstm.to_csv(arquivo_submissao, index=False)
print(f"Arquivo salvo: {arquivo_submissao}")
print(f"Linhas: {len(df_submission_convlstm):,}")
display(df_submission_convlstm.head())
display(df_submission_convlstm.tail())